lighthouse-core/audits/legacy-javascript.js

```diff
@@ -0,0 +1,334 @@
+    let result;
+    let line = 0;
+    let lineBeginsAtIndex = 0;
+    // Each pattern maps to one subgroup in the generated regex. For each iteration of RegExp.exec,
+    // only one subgroup will be defined. Exec until no more matches.
+    this.re.lastIndex = 0;
+    while ((result = this.re.exec(code)) !== null) {
+      // Discard first - it's the whole matching pattern.
+      // Index 1 is truthy if matching a newline, and is used to track the line number.
+      // `patternExpressionMatches` maps to each possible pattern.
+      // Only one of [isNewline, ...patternExpressionMatches] is ever defined.
+      const [, isNewline, ...patternExpressionMatches] = result;
+      if (isNewline) {
+        line++;
+        lineBeginsAtIndex = result.index + 1;
+        continue;
+      }
+      const pattern = this.patterns[patternExpressionMatches.findIndex(Boolean)];
+
+      // Don't report more than one instance of a pattern for this code.
+      // Would result in multiple matches for the same pattern, ex: if both '='
+      // and 'Object.defineProperty' are used conditionally based on feature detection.
+      // Would also result in many matches for transform patterns.
+      if (seen.has(pattern)) {
+        continue;
+      }
+      seen.add(pattern);
+
+      matches.push({
+        name: pattern.name,
+        line,
+        column: result.index - lineBeginsAtIndex,
+      });
+    }
+
+    return matches;
+  }
+}
+
+class LegacyJavascript extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'legacy-javascript',
+      scoreDisplayMode: Audit.SCORING_MODES.INFORMATIVE,
+      description: str_(UIStrings.description),
+      title: str_(UIStrings.title),
+      requiredArtifacts: ['devtoolsLogs', 'ScriptElements'],
+    };
+  }
+
+  /**
+   * @param {string?} object
+   * @param {string} property
+   */
+  static buildPolyfillExpression(object, property) {
+    const qt = (/** @type {string} */ token) =>
+      `['"]${token}['"]`; // don't worry about matching string delims
+
+    let expression = '';
+
+    // String.prototype.startsWith =
+    expression += `${object || ''}\\.?${property}\\s*=`;
+
+    // String.prototype['startsWith'] =
+    expression += `|${object || ''}\\[${qt(property)}\\]\\s*=`;
+
+    // Object.defineProperty(String.prototype, 'startsWith'
+    expression += `|defineProperty\\(${object || 'window'},\\s*${qt(property)}`;
+
+    if (object) {
+      const objectWithoutPrototype = object.replace('.prototype', '');
+      // e(e.S,"Object",{values
+      // minified pattern found in babel-polyfill
+      // see https://cdnjs.cloudflare.com/ajax/libs/babel-polyfill/7.2.5/polyfill.min.js
+      expression += `|;e\\([^,]+,${qt(objectWithoutPrototype)},{${property}`;
+    }
+
+    return expression;
+  }
+
+  /**
+   * @return {Pattern[]}
+   */
+  static getPolyfillPatterns() {
+    return [
+      'Array.prototype.filter',
```

Comment:
this is everything from that spreadsheet, minus ones that don't fit the model (the `*[Symbol.*]` ones) and the ones I figured were too much (`Math.*` mostly)